core/trino-main/src/main/java/io/trino/execution/scheduler/UniformNodeSelector.java

```diff
@@ -399,4 +410,64 @@ private static boolean isSplitLocal(List<HostAddress> splitAddresses, HostAddres
         }
         return false;
     }
+
+    private static class QueueSizeAdjuster
+    {
+        private final Map<String, Long> splitsWeightPerNode;
+        private final Set<String> previousScheduleFullNodes;
+        private final Map<String, Long> previouslyScheduledWeight;
+        private final long minPendingSplitsWeightPerTask;
+        private final long maxPendingSplitsWeightPerTask;
+        private final long maxSplitsWeightPerNode;
+
+        QueueSizeAdjuster(long minPendingSplitsWeightPerTask, long maxPendingSplitsWeightPerTask, long maxSplitsWeightPerNode)
+        {
+            this.splitsWeightPerNode = new HashMap<>();
+            this.previousScheduleFullNodes = new HashSet<>();
+            this.previouslyScheduledWeight = new HashMap();
+            this.maxPendingSplitsWeightPerTask = maxPendingSplitsWeightPerTask;
+            this.minPendingSplitsWeightPerTask = minPendingSplitsWeightPerTask;
+            this.maxSplitsWeightPerNode = maxSplitsWeightPerNode;
+        }
+
+        public void update(List<RemoteTask> existingTasks, NodeAssignmentStats nodeAssignmentStats)
+        {
+            if (maxPendingSplitsWeightPerTask == minPendingSplitsWeightPerTask) {
+                return;
+            }
+            for (RemoteTask task : existingTasks) {
+                String nodeId = task.getNodeId();
+                long adjustedMaxSplitsWeightPerTask = getAdjustedMaxPendingSplitsWeightPerTask(nodeId);
+
+                if (adjustedMaxSplitsWeightPerTask < maxPendingSplitsWeightPerTask
+                        && previousScheduleFullNodes.contains(nodeId)
```

Comment:
why do we need `previousScheduleFullNodes.contains(nodeId)` ?

Reply:
If a node was not exhausted during the previous iteration, we don't want to extend its queue.

Reply:
why not? I would add a comment here or in some place where this algorithm is described.